### PR TITLE
Fix missing update action for git deps when shrinkwrap is used

### PR DESCRIPTION
**Pull request by kossnocorp**
Prevent inflateShrinkwrap function from relying on a package version for git dependencies. That was the cause of two problems:
1. When node_modules dir contains an outdated git dependency, it won't be updated during npm install.
2. npm shrinkwrap command fails.

For more details about the bug, see https://github.com/npm/npm/issues/12718.

_I still working on tests, will appreciate some guidance on that._

### CODE REVIEW

**Comment by kossnocorp** [comment]
@iarna @othiym23 hey! Is there any chance to get this reviewed and merged?

**Comment by iarna** [comment]
@kossnocorp We evaluate PRs for inclusion sometime between Mon and Wed for inclusion the following Thursday.

File: lib/install/inflate-shrinkwrap.js [hunk 1]
```diff
@@ -29,7 +29,7 @@ var inflateShrinkwrap = module.exports = function (tree, swdeps, finishInflating
     if (child && (child.fromShrinkwrap ||
                   (sw.resolved && child.package._resolved === sw.resolved) ||
                   (sw.from && url.parse(sw.from).protocol && child.package._from === sw.from) ||
-                  child.package.version === sw.version)) {
+                  (!child.package.gitHead && child.package.version === sw.version))) {
```
Comment (iarna) [inline lib/install/inflate-shrinkwrap.js]: I'd like a more general correction here. This code is saying that a module on disk matches the shrinkwrap if ANY of these are true, and I'm pretty sure that's not actually what we want to do.

I don't know how I got my head screwed on wrong when I wrote that, but I'm ​_pretty sure_​ I intended those to be exclusive, eg, if there's a _resolved then rely on that for identity, else if there's a from with a protocol, use that for identity, and if nothing else is available use version for identity.

So I _think_ this should actually read:
```js
    if (child) {
      var useExisting
      if (child.fromShrinkwrap) {
        useExisting = true
      } else if (sw.resolved) {
        useExisting = child.package._resolved === sw.resolved
      } else if (sw.from && url.parse(sw.from).protocol) {
        useExisting = child.package._from === sw.from
      } else if (child.package.version === sw.version) {
        useExisting = true
      }
      if (useExisting) {
        if (!child.fromShrinkwrap) child.fromShrinkwrap = spec
        tree.children.push(child)
        return inflateShrinkwrap(child, sw.dependencies || {}, next)
      }
    }
```

**Comment by iarna** [comment]
Regarding the test, we have been working hard to eliminate any dependencies that talk to the network, so I would much appreciate it if you could rework this to do its work locally.

You can see an example of creating a little git repo and firing up a local git server in [`add-remote-git.js`](https://github.com/npm/npm/tree/master/test/tap/add-remote.git.js), though I'd prefer new tests to write (and cleanup) their files with [`tacks`](npmjs.com/package/tacks) (see [`gently-rm-linked-module.js`](https://github.com/npm/npm/tree/master/test/tap/gently-rm-linked-module.js) as an example) instead of doing it by hand.

**Comment by kossnocorp** [comment]
@iarna thank you, working on that 👍

**Comment by kossnocorp** [comment]
@iarna sorry for the delay, here is the change log:

1. Rebase the branch onto master
2. Apply your solution with refactored `childIsEquivalent` function (kudos for flawless GitHub-comment-programming!)
3. Make tests work offline.

I've tried Tacks approach, but frankly code got even messier because I still have to manually handle a lot of operations in order to reproduce the bug.

**Comment by kossnocorp** [comment]
😉

**Comment by kossnocorp** [comment]
❤️

**Comment by conf** [comment]
@iarna Any update on this? This is a blocker for using `npm-shrinkwrap` feature for us since we depend on some package forks on Github and it's a pain to update for everybody in the team.

**Comment by kossnocorp** [comment]
😿

**Comment by iarna** [comment]
This is merged and will be in the next release (scheduled for later today).

**Comment by kossnocorp** [comment]
🙏

**Pull request closed without merging**